noether_filtering/include/noether_filtering/impl/filter_group.hpp

```diff
@@ -0,0 +1,201 @@
+  }
+  catch (const XmlRpc::XmlRpcException &ex)
+  {
+    CONSOLE_BRIDGE_logWarn("Failed to parse '%s' parameter: '%s'", CONTINUE_ON_FAILURE.c_str(), ex.getMessage().c_str());
+    continue_on_failure_ = false;
+  }
+
+  // load filter infos
+  std::vector<FilterInfo> filter_infos;
+  XmlRpc::XmlRpcValue filter_configs = config[FILTERS];
+  if(!loadFilterInfos(filter_configs, filter_infos))
+  {
+    CONSOLE_BRIDGE_logError("Failed to load filters");
+    return false;
+  }
+
+  // instantiating filters
+  for (FilterInfo &fi : filter_infos)
+  {
+    if (filters_map_.count(fi.name) > 0)
+    {
+      CONSOLE_BRIDGE_logError("The %s plugin '%s' has already been added",
+                              utils::getClassName<F>().c_str(),
+                              fi.name.c_str());
+      return false;
+    }
+
+    FilterBasePtr plugin;
+    try
+    {
+      //plugin = filter_loader_.createUniqueInstance<FilterBase<F>>(fi.type_name);
+      plugin.reset(filter_loader_.createUnmanagedInstance(fi.type_name));
+      if(!plugin->configure(fi.config))
+      {
+        CONSOLE_BRIDGE_logError("%s plugin '%s' failed to load configuration",utils::getClassName<F>().c_str(),
+                  fi.name.c_str());
+        return false;
+      }
+
+      // storing plugin
+      filters_loaded_.push_back(fi.name);
+      filters_map_.insert(std::make_pair(fi.name, std::move(plugin)));
+    }
+    catch(const class_loader::ClassLoaderException& ex)
+    {
+      CONSOLE_BRIDGE_logError("%s plugin '%s' could not be created",utils::getClassName<F>().c_str(),
+                fi.name.c_str());
+      return false;
+    }
+  }
+
+  return true;
+}
+
+template<class F>
+bool FilterGroup<F>::applyFilters(const F& input, F& output, std::string& err_msg)
+{
+  // applying all loaded filters
+  return applyFilters(filters_loaded_,input, output, err_msg);
+}
+
+template<class F>
+bool FilterGroup<F>::applyFilters(const std::vector<std::string>& filters, const F& input, F& output,
+                                    std::string& err_msg)
+{
+  std::vector<std::string> selected_filters = filters;
+  if(selected_filters.empty())
+  {
+    CONSOLE_BRIDGE_logWarn("%s received empty list of filters, using all filters loaded", utils::getClassName<FilterGroup<F>>().c_str());
+    std::copy(filters_loaded_.begin(), filters_loaded_.end(), std::back_inserter(selected_filters));
+  }
+
+  // check all filters exists
+  auto check_fn = [this, &err_msg](const std::string &f) {
+    if (filters_map_.count(f) == 0)
+    {
+      err_msg = boost::str(boost::format("The filter %s was not found") % f);
+      CONSOLE_BRIDGE_logError("%s", err_msg.c_str());
+      return false;
+    }
+    return true;
+  };
+
+  if(!std::all_of(filters.begin(), filters.end(), check_fn))
+  {
+    return false;
+  }
+
+  // apply filters
+  bool success = false;
+  for(const std::string& fname : selected_filters)
+  {
+    bool current_filter_succeeded = filters_map_.at(fname)->filter(input,output);
```

Comment:
This doesn't actually filter with more than one filter in the chain. It just uses the original input for every filter. We need to add a temporary "input" that gets overwritten by the output at the end of the filter stage.

Reply:
Good catch

Reply:
I made a fix for this in #58